---
title: "Analyze an *A.thaliana* simulated phenotype"
author: "Héctor Climente-González"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Running SConES

In this example we will analyze GWAS data from *Arabidopsis thaliana*. (reference) This data is publicly available from [easyGWAS](https://easygwas.ethz.ch/):

```{r fetch data, eval=FALSE}
tmp <- tempfile()
download.file("https://easygwas.ethz.ch/down/dataset/download/1/",tmp)
unzip(tmp)
unlink(tmp)
```

We start by loading the `martini`. 

```{r load martini}
library(martini)
```

`martini` imports the package `snpStats`. We will use its functions to read the GWAS data.


```{r read data, eval=FALSE}
library(snpStats)
gwas <- read.pedfile("genotype.ped", snps = "genotype.map")
```

Other than the GWAS data, we need a network of SNPs to work on. We can create those with the family of `get_G*_network` function. In this vignette, we will work with a GM network, a network where SNPs are connected if they belong to the same gene. We provide the `get_GM_network` function with the SNP data and the tax ID of *Arabidopsis thaliana*. Sit comfortably, this step might take a while.

```{r create GM network, eval=FALSE}
net <- get_GM_network(gwas, 3702)
```

The dataset provides several phenotypes in `phenotype.pheno`. However, most of these phenotypes were only measured for a small subset of the subjects. In consequence, in this example we will work with a simulated phenotype. First we will simulate 60 causal SNPs taking 60 SNPs tightly interconnected in the GM network.

```{r simulate causal SNPs, eval=FALSE}
causal <- simulate_causal_snps(net, 60)
```

We will use these SNPs to simulate a qualitative phenotype. Other than the desired number of cases and the number of controls, we pick the heritability of the disease. In this example we work with a fully heritable phenotype.

```{r simulate phenotypes, eval=FALSE}
y <- simulate_phenotype(gwas, causal, h2=1, qualitative=T, ncases=653, ncontrols=654)
```

We need to overwritte the original phenotype column from the PED file with the new phenotype.

```{r overwrite affected, eval=FALSE}
gwas$fam$affected <- y
```

We can proceed to search for cones (CONnected Explanatory SNPs). We will choose chi-squared as association score, as we are working with a qualitative genotype.

```{r search cones, eval=FALSE}
cones <- search_cones(athal$gwas, athal$gm, associationScore = "chi2")
```

```{r check overlap}
data(athal)
cones <- athal$cones
cones$causal <- cones$snp %in% names(athal$causal)
table(subset(cones, select = c("selected","causal")))
```
We recover all causal 60 SNPs, and 129 false positives. Let's see why they were selected.

```{r manhattan}
library(tidyverse)

filter(cones, selected) %>%
  ggplot(aes(y = c, x = ifelse(causal, "Yes", "No"), fill = causal)) +
  geom_violin(alpha = 0.5) +
  geom_jitter(position = position_jitter(width = 0.1, height = 0),
              alpha = 0.3) +
  scale_fill_manual(values=c("TRUE"="firebrick", "FALSE"="gray")) +
  labs(x = "Causal", y = "Association score") +
  theme(legend.position="none")
```

A significant proportion of the causal SNPs do not have a strong association with the phenotype. 

```{r subnetwork}
library(igraph)

selected <- subset(cones,selected)
selected$snp <- as.character(selected$snp)

g <- induced.subgraph(graph = athal$gm, selected$snp)
g <- set_vertex_attr(g, "color", index = match(selected$snp, V(g)$name), selected$causal)
V(g)$color <- ifelse(V(g)$color, "firebrick", "grey")

plot.igraph(g, vertex.label = NA, layout = layout.fruchterman.reingold)
```

Although promising, maybe we can find some better hyperparameters.

```{r}
cones <- search_cones(athal$gwas, athal$gm, associationScore = "chi2", debug = T, etas = c(10,20,50,100,200,300), lambdas = c(10,20,50,100,200,300))
```

[0;31mModel score matrix[0m
[0;31m       0       10       20       50      100      200      300
      10     -inf     -inf     -inf     -inf     -inf     -inf
      20    -6303 -5578.44 -4810.49  -4566.5 -3598.19 -4091.36
      50 -1825.39 -1825.87 -1782.43 -1838.32 -2006.54 -2006.54
     100 -1382.92 -1425.42  -1376.2  -1376.2  -1376.2 -1343.19
     200 -1583.41 -1583.41   -1e+31   -1e+31   -1e+31   -1e+31
     300   -1e+31   -1e+31   -1e+31   -1e+31   -1e+31   -1e+31[0m
[0;31mAverage number of selected SNPs.[0m
[0;31m      0      10      20      50     100     200     300
     10 10433.9  8635.9  6405.2  4970.6    3601  2746.8
     20  1524.6  1249.7   941.6   802.3   658.3   644.6
     50   162.9   160.8   158.8   163.4     176     173
    100    82.5      77      67    57.2    57.2    54.5
    200     2.2     1.1       0       0       0       0
    300       0       0       0       0       0       0[0m
