---
title: "Using SConES to analyze an *A.thaliana* simulated phenotype"
author: "Héctor Climente-González"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r load martini, warning = FALSE}
library(martini)
```

```{r athaliana generation, eval = FALSE}
library(martini)

# fetch the data
tmp <- tempfile()
download.file("https://easygwas.ethz.ch/down/dataset/download/1/",tmp)
unzip(tmp)
unlink(tmp)

# read gwas experiment
gwas <- read.pedfile("genotype.ped", snps = "genotype.map")

# create GM network
net <- get_GM_network(gwas, 3702)

# generate simulated phenotypes
causal <- simulate_causal_snps(net, 60)
y <- simulate_phenotype(gwas, causal, h2=1, ncases=653, ncontrols=654, qualitative=T)

## write back into the SnpMatrix object
gwas$fam$affected <- y
```
