---
title: "Run SConES on a simple dataset"
author: "Héctor Climente-González"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Running SConES

In this example we will analyze a simple case of GWAS data. We start by loading the `martini`, which contains the example we will work on.

```{r load martini}
library(martini)
```

`minigwas` is a simplification of an actual GWAS. It is a list equivalent to what `snpStats::read.pedfile` would generate. In this experiment, we sequenced 25 SNPs on 100 samples. Half of them suffer from a disease, the other half are perfectly healthy individuals. The information of each patient is available in the `fam` element:

```{r print fam}
head(minigwas$fam)
```

We pay close attention to the `affected` column, with takes value `2` if the patient is a case, or `1` if they are control. 

The other important element is `map`, which contains the SNP information:

```{r print map}
head(minigwas$map)
```

In `minigwas`, the SNP name is a three character word, where every letter has a meaning: the first character indicates the chromosome, the second the gene, and the third its position inside the gene. 

Lastly, `genotypes` contains the actual genotype for each SNP and sample.

```{r print genotypes}
minigwas$genotypes
```

Other than the GWAS data, we need a network of SNPs to work on. We can create those with the family of `get_G*_network` function. Let's see how they work.

```{r create GS network}
gs <- get_GS_network(minigwas)
par(mar=c(0,0,0,0)+.1)
plot(gs)
```

```{r create GM network}
gm <- get_GM_network(minigwas, snpMapping = minisnpMapping)
par(mar=c(0,0,0,0)+.1)
plot(gm)
```

```{r create GI network}
gi <- get_GI_network(minigwas, snpMapping = minisnpMapping, ppi = minippi)
par(mar=c(0,0,0,0)+.1)
plot(gi)
```

```{r find causal SNPs}
cones <- search_cones(minigwas, gi)
head(cones)
```

```{r manhattan plot}
library(tidyverse)

ggplot(cones, aes(x = pos, y = c, color = ifelse(selected, "Yes", "No"))) +
  geom_point() +
  facet_grid(. ~ chr) +
  labs(x = "Genomic position", y = "Association score", color = "Selected") +
  theme_minimal() +
  theme(legend.position = "bottom")
```